WIP: Ben sepanski/message formatter [VAN-12]

benSepanski opened the pull request:
This implements a `ReportFormatter` object that handles most of the mundane issues for formatting by allowing developers to declare intent (begin section, list, format a function, etc.). Actual formatting is configured via the `FormatSettings` object.

The report formatter is then used by the Reentrancy Detector

Hopefully we can reuse this for multiple detectors, so feedback is appreciated!

benSepanski commented:
@0xddom you mentioned something like this in the slack, is this close to what you were wanting?

0xddom commented:
> @0xddom you mentioned something like this in the slack, is this close to what you were wanting?

Not exactly but I think this could be pretty useful too. What I had in mind was a builder that removes the cumbersomeness of smart pointers, leaving a more terse and readable statement for creating sarif objects. (see the latest commit in the sarif branch for an example: https://github.com/Veridise/Vanguard/blob/dani/sarif-support/lib/detectors/FunctionPrinter/FunctionPrinter.h#L23). The idea is to make it a simple as possible to motivate people into using it.

This formatters would return a string in the end, no? The we could just plug the end of the formatter to the sarif result class (via the `DetectorResult` for example). Then the formatter would just handle printing to two places: stdout and the string that stores the message in the sarif "world".

benSepanski commented:
> Not exactly but I think this could be pretty useful too. What I had in mind was a builder that removes the cumbersomeness of smart pointers, leaving a more terse and readable statement for creating sarif objects. (see the latest commit in the sarif branch for an example: https://github.com/Veridise/Vanguard/blob/dani/sarif-support/lib/detectors/FunctionPrinter/FunctionPrinter.h#L23). The idea is to make it a simple as possible to motivate people into using it.

I see, so mainly adding a `toSARIF`-type function to the report formatter?

> This formatters would return a string in the end, no? The we could just plug the end of the formatter to the sarif result class (via the `DetectorResult` for example). Then the formatter would just handle printing to two places: stdout and the string that stores the message in the sarif "world".

Yes, though eventually the formatting functions which write `Function` and `Variable` names can automatically record their logical locations for output to SARIF behind the scenes. That way the user shouldn't have to touch SARIF locations at all

0xddom commented:
>I see, so mainly adding a toSARIF-type function to the report formatter?

> Yes, though eventually the formatting functions which write Function and Variable names can automatically record their logical locations for output to SARIF behind the scenes. That way the user shouldn't have to touch SARIF locations at all

Would depend on how much information is encapsulated in the formatter. Each detector is meant to implement its own formatter?

What I like the most is to have the `DetectorResult` which holds one of the results a detector can yield. That `DetectorResult` knows how to express the result in SARIF because it implements the `toSARIFResult`. Then the `DetectorResult` uses the `Formatter` to prepare the raw data for crafting the SARIF `Result` object.

This way we:
- Decouple the formatters and SARIF
- If the ceremony for taking data from the formatter and writing it into sarif is general we can use the same `DetectorResult` class in all detectors instead of implementing each their own. This requires that detectors now implement a custom `Formatter` I guess.
- The responsibility of the formatter would be to prepare data, but not actually writing it anywhere.

0xddom commented:
Looking a bit more into your code I think I see a difference. The formatter you are using in the reentrancy check seems to generate a whole document, while the idea behind SARIF is to generate `Result` objects that are stored in the JSON. So, I have another idea:

Implement this formatter idea as a SARIF output format, removing completely the presentation layer from the detector classes. The SARIF module supports creating new dumpers aside from JSON, we could leverage that to generate plain text reports using the SARIF data model directly

benSepanski commented:
>Looking a bit more into your code I think I see a difference. The formatter you are using in the reentrancy check seems to generate a whole document, while the idea behind SARIF is to generate Result objects that are stored in the JSON. So, I have another idea:

I thought the SARIF report was one big object, but I guess it is one report per "bug"?

>Implement this formatter idea as a SARIF output format, removing completely the presentation layer from the detector classes. The SARIF module supports creating new dumpers aside from JSON, we could leverage that to generate plain text reports using the SARIF data model directly

I guess my thinking was that the developer would want a lot of control over how the text formatting was derived, but given that description it would be relatively easy to generate the SARIF reports automatically. This way the SARIF report messages could include the text formatting as well, e.g. to standardize function names, lists, etc. in the SARIF messages. Does SARIF have a nice way to represent these things?

Side note--does this entail using the unique pointers you were unhappy with?

0xddom commented:
>I thought the SARIF report was one big object, but I guess it is one report per "bug"?

Yes, there is a "larger" object but is outside of the detector's scope that just collects all the results and writes the final report.

>I guess my thinking was that the developer would want a lot of control over how the text formatting was derived, but given that description it would be relatively easy to generate the SARIF reports automatically. This way the SARIF report messages could include the text formatting as well, e.g. to standardize function names, lists, etc. in the SARIF messages. 

The detectors author wants control over the formatting of each report yes, that makes total sense. According to the spec you can write results in plain text, markdown, etc. We can leave that to the author. Then to generate a full plain text report we can use the SARIF dumping API. So we can implement nice formatters like the ones you are writing that write the text description of the result and then feed that to the SARIF result.

>Does SARIF have a nice way to represent these things?

No, if we want to standardise those things (Functions, Vars, etc) we would need a class model for that.

>Side note--does this entail using the unique pointers you were unhappy with?

Ah, no, no. Not unhappy. I love them, but every time I have to type out `std::unique_ptr<Something something>` a bit of my soul dies, so every piece of API that I can write that automatically infers or adds that is great. We should use smart pointers as much as possible, specially in this modules that move data around.

benSepanski commented on lib/detectors/CallGraphReentrancy/PossibleReentrancy.h:
```diff
@@ -120,145 +105,39 @@ namespace vanguard {
-            return report;
+        [[nodiscard]] const Function *getReentryPointFunction() const {
+            assert(!reentryPointSet.empty());
+            return (*reentryPointSet.begin())->block()->function();
```
Yes, [mergeFrom](https://github.com/Veridise/Vanguard/blob/ca4d1d1a4b76d8acc96be34270f30184696ff725/lib/detectors/CallGraphReentrancy/PossibleReentrancy.h#L93-L96) enforces this 

This is mainly just to reduce verbosity in the report. During detection, each reentry point is handled separately. However, since reentry points in the same function usually have very similar "reachable modification sets," it's helpful to merge them into a single report.

benSepanski commented on lib/detectors/formatters/FormatSettings.h:
```diff
@@ -0,0 +1,177 @@
+            std::string titleBar = titleBarStream.str();
+
+            std::string spaces(titleBarStr.size(), ' ');
+            llvm::outs() << titleBar << "\n"
```
Ack--yes. Thanks!

benSepanski commented on lib/detectors/formatters/FormatSettings.h:
```diff
@@ -0,0 +1,177 @@
+        ListSettings list{};
+        MetadataSettings metadata{};
+        SectionSettings section{};
+        VariableSettings<Domain> variable{};
```
Good point, fixed in https://github.com/Veridise/Vanguard/pull/88/commits/530b751f8c87829309c3f8c671ffd75967ec692a

benSepanski commented on lib/detectors/formatters/FormatSettings.h:
```diff
@@ -0,0 +1,177 @@
+namespace vanguard::formatter {
+
+    template<typename PrintedObj>
+    class CustomPrinter {
```
Fixed in https://github.com/Veridise/Vanguard/pull/88/commits/46b118e2cad6130bc63b7535e02ce201763331b4

benSepanski commented on lib/output/formats/Format.h:
```diff
@@ -0,0 +1,21 @@
+    template<class O>
+    class Format {
+    public:
+        virtual O format(Report report) = 0;
```
I think we're missing a virtual destructor here

benSepanski commented on lib/output/Issue.h:
```diff
@@ -0,0 +1,46 @@
+#include "Localizable.h"
+#include "detectors/Detector.h"
+
+namespace vanguard {
```
Can we put all the formatting stuff in a `vanguard::formatter` namespace, rather than just `vanguard`?

benSepanski reviewed (commented):
Looks great! Thanks for doing all this :)

benSepanski commented on lib/output/Issue.h:
```diff
@@ -8,8 +8,10 @@
+#include "Build.h"
 
-namespace vanguard {
+namespace vanguard::formatting {
```
Oh I see, this is designed to match SARIF

0xddom replied:
yeah, that was the inspiration, but we could have more

0xddom commented on lib/detectors/Detector.h:
```diff
@@ -33,6 +33,8 @@ namespace vanguard {
             collectedResults.push_back(std::move(result));
         }
+
+        virtual std::string getName() const = 0;
```
wow! that looks cool

Technius replied:
What's wrong with a virtual method here? CRTP is powerful but has a debugging experience that is awfully equal in magnitude.

0xddom replied:
The main issue of the virtual method would be that we use the static one for populating the registry and having a non static one would require to instantiate each detector for checking the name (I think).

Technius replied:
CRTP makes a lot more sense in this case, thanks. We could also add a helper function `getDetectorName<Detector>()` that calls `Detector::name()`

0xddom replied:
one question. If we have the `getName` virtual method in `Detector` which is where the reporting API expects it according to the interface, would the implementation in the mixin count? We cannot expect a `Named` because that type depends on the concrete detector type

benSepanski replied:
> The main issue of the virtual method would be that we use the static one for populating the registry and having a non static one would require to instantiate each detector for checking the name (I think).

Yes that's my issue with it. Having to add
```
std::string getName() override {
    name()
}
```
to every detector bugs me.

> CRTP makes a lot more sense in this case, thanks. We could also add a helper function `getDetectorName<Detector>()` that calls `Detector::name()`

Unless I'm misunderstanding, this may not work for all the cases we need. We have some object `detector` and want to get its name, without knowing its concrete type at compile time

> one question. If we have the getName virtual method in Detector which is where the reporting API expects it according to the interface, would the implementation in the mixin count? We cannot expect a Named because that type depends on the concrete detector type

Haha, this is a good question. Turns out I was tricked by chatGPT and forgot about the diamond inheritance problem. My CRTP answer doesn't actually work :(

0xddom replied:
Good to know that our jobs are safe from our AI overlords for now! Back to the drawing board on this one I guess

benSepanski replied:
Haha :)

One slightly ugly option is to allow for a `Named<B,A>` which creates the inheritance graph:
```
B : Named<B,A> : A : Named<A>
```
For instance, the following would work (I've actually checked this one now haha)
```
/// Dummy class for base case
class NamedBase{};

template<typename Derived, typename Base = NamedBase>
class Named : public Base {
public:
    virtual std::string getName() const {
        return Derived::name();
    }
};

class A : public Named<A> {
public:
    static std::string name() {
        return "A";
    }
};

class B : public Named<B,A> {
public:
    static std::string name() {
        return "B";
    }
};

int main()
{
    A *b = new B();
    std::cout << b->getName();

    return 0;
}
```

0xddom replied:
Thats code that only a mother could love... Damn... What if we implement all that in `Detector` instead of `Named`?

benSepanski replied:
Lol that's a perfect way to put it. Good idea though, maybe something like

```
class BaseNameResolver {
public:
    virtual std::string getName() const {
        assert(false);
    }
};

template<typename Derived>
class NameResolver : public BaseNameResolver {
public:
    virtual std::string getName() const {
        return Derived::name();
    }
};

class A  {
public:
    template<typename NamedType>
    A(NameResolver<NamedType> *n) {
        nameResolver = n;
    }
    
    A() {
        nameResolver = new NameResolver<A>;
    }
    
    static std::string name() {
        return "A";
    }
    
    std::string getName() {
        return nameResolver->getName();
    }

protected:
    BaseNameResolver *nameResolver;
};

class B : public A {
public:
    B() : A::A(new NameResolver<B>) {}

    static std::string name() {
        return "B";
    }
};
```

0xddom replied:
whats `A` and `B` here? `Detector` and `AnArbitraryDetector`?

benSepanski replied:
From vanguard meeting:
* Rename static function as "cmdLineName"
* Keep virtual "name" function

0xddom commented on lib/output/AtomicObject.h:
```diff
@@ -0,0 +1,20 @@
+        };
+
+    } // vanguard
+} // formatting
```
those were machine generated. CLion was having a normal one...

0xddom commented on lib/output/Build.h:
```diff
@@ -0,0 +1,44 @@
+    class ListBuilder;
+    class ParagraphBuilder;
+
+    class TextBuilder {
```
Yeah, we need to discuss that part because we most likely will have to do templates but I would rather avoid them

benSepanski replied:
Just realized that one fun option is to template the methods individually!

I think this will work as something like:
```
template<typename Domain>
operator<<(typename Domain::Function *f) {
    this->append(formatter::Function(f));
}
```
I'll have to double-check if template argument deduction will work in this case though....

0xddom replied:
what happens if the domain does not have the object? Can we check at compile time? `CircuitDomain` does not have a `Contract` type so if I try to do `Domain::Contract` and then instantiate `Domain` with `CircuitDomain` I get a type error

benSepanski replied:
The template is instantiated lazily, so this shouldn't be an issue. Basically, the function only exists if you call it

I think the main issue will be argument deduction, otherwise we have to supply `Domain` to the `<<` operator

0xddom commented on lib/output/formats/SARIFFormat.cpp:
```diff
@@ -3,3 +3,42 @@
+#include <memory>
+
+namespace vanguard::formatting {
+    SARIF::Level severity2level(Severity severity) {
```
sure, I dont have a strong preference

Technius commented on lib/output/AtomicString.cpp:
```diff
@@ -0,0 +1,10 @@
+namespace vanguard {
+    namespace formatting {
+    } // vanguard
+} // formatting
```
We should set up clang-format so we don't need to worry about stuff like this

0xddom replied:
agree

0xddom commented on lib/output/Paragraph.h:
```diff
@@ -0,0 +1,24 @@
+        class Paragraph : public CompositeObject {
+        public:
+            Paragraph();
+            std::shared_ptr<CompositeObject> getTitle();
```
yeah, that thought crossed my mind. I would rather do the latter

Technius commented on lib/output/Issue.cpp:
```diff
@@ -4,5 +4,62 @@
+    IssueBuilder::IssueBuilder() : TextBuilder(std::make_shared<CompositeObject>()) {}
+
+    std::unique_ptr<Issue> IssueBuilder::toIssue() {
+        return std::make_unique<Issue>(properties, nullptr);
```
Note that we're building in release mode right now, so asserts will be ignored. It would be better to use a pattern like `std::cerr << "my message\n"; std::abort()` for the moment.

0xddom replied:
good point

0xddom replied:
is there something like rust's `todo!()` in c++ that is quick and easy to type? you know, so that my smooth brain actually remembers to use it

Technius replied:
Unfortunately, no built in solution. I was thinking we could add a macro in a utility/helper header.
```c++
#define VG_ASSERT(cond)                                                        \
    if (!cond) {                                                               \
        std::cerr << "Assertion failed: " #cond ", " __FILE__ ":" << __LINE__  \
                  << std::endl;                                                \
        std::abort();                                                          \
    }
```

Then you can define a `VG_TODO` macro as `VG_ASSERT(false && "todo")`.

0xddom replied:
if Jon does not kill us for using macros that looks comfy enough

benSepanski replied:
I am pro `VG_ASSERT` macro ❤️

0xddom commented on lib/output/Issue.h:
```diff
@@ -24,15 +26,24 @@ namespace vanguard {
 
     class Issue {
+    public:
+        Issue(IssueProperties properties, std::shared_ptr<FormattedObject> message);
```
the setter methods are in the builder, which fits the pattern

0xddom commented on lib/output/formats/Format.h:
```diff
@@ -5,7 +5,9 @@
-namespace vanguard {
+#include "../Report.h"
+
+namespace vanguard::formatting {
```
yes, a reference makes sense

benSepanski commented:
Pushing for now to publish my progress/ensure it doesn't get lost.

I've implemented most of the "Document Tree" construction

I've also adjusted the Reentrancy Detector to use the new framework, which will be convenient for testing (since we should be able to reproduce the same output for the detector)

The main thing left to do is implement the `PlainTextFormat::format` function, which will involve porting over my code from the previous formatter. I will do this tomorrow

0xddom commented:
Awesome! Wrt the `*Format` classes, I think we should have two methods: One for generating the output of a full report and another one for generating an arbitrary formatter object, that way we can leverage the `PlainTextFormat` class for sarif to generate what are string types like the message of a result.

Kind of like:

```
class Format<O> {
  O format(Report report);
  std::string format(FormattedObject obj);
}
```

benSepanski commented:
Yeah I like that

0xddom commented:
CI is broken because some detectors in `v1` do not adhere to the new interface and they do not exist in our branch. Nothing to worry for now

benSepanski commented:
@0xddom I've just pushed the major changes. I am still debugging three of the reentrancy tests, and then have to worry about some whitespace issues (which are not caught by the tests). After that I can help write some more unit tests

Feel free to make a pass over my changes  now, future changes should be pretty minor

benSepanski commented:
I've now reproduced the desired reentrancy detector output 😄 

I think the remaining things to do here are
- [ ] Have @0xddom review my changes and address fixes
- [ ] Write some more unit tests for the formatter specifically (many things like whitespace/empty lines aren't caught by standard lit tests, so we'll want some that check for those things)

Then we're ready for third-party review?

benSepanski commented:
I've added some basic unit tests for lists, paragraphs, and nested lists. These tests check basic whitespace/indentation properties.

Since the detector lit tests will be handling all the Localizable objects (e.g. outputting functions, etc.) I figured it was less essential to spend the effort of mocking functions for unit tests

If there's some other area you want me to provide more coverage on, let me know. Otherwise, I'm going to leave it at this for now

0xddom commented:
LGTM

Next steps I think are:

- [ ] Add Domain specific wrappers in the same fashion as `Function` and `Variable`.
- [ ] Fix CI
- [ ] Integrate the new API into the detectors (We said not to do it in the PR, right?)

The main core stuff I think is stable enough for a code review since this changes are either outside the module or small additions

Technius commented:
This is getting rather large. Let's get an initial version merged in so we can start using it and to make future reviews easier.

benSepanski commented:
So @Technius I should go ahead and merge?

Technius commented:
I'm making a final review pass, give me until EOD today.

benSepanski commented:
Sgtm. I'll probably be out of office by then so feel free to merge if you're happy with it

Technius commented on lib/output/Issue.h:
```diff
@@ -0,0 +1,62 @@
+        IssueBuilder();
+        TextBuilder title();
+        std::unique_ptr<Issue> toIssue();
+        IssueBuilder *setSeverityLevel(Severity severity);
+        IssueBuilder *setInfo();
+        IssueBuilder *setWarn();
+        IssueBuilder *setCritical();
+        IssueBuilder *setCodeLocation(const Localizable *localizable);
+        IssueBuilder *setFunctionLocation(const Localizable * localizable);
+        IssueBuilder *setModule(const Localizable * localizable);
+        IssueBuilder *setDetector(const Detector *detector);
```
nit: might be better to make each return value a reference to emphasize that it is non-null.

Technius commented on lib/output/Localizable.h:
```diff
@@ -0,0 +1,32 @@
+    class Localizable {
+    public:
+        virtual std::optional<std::string> locationName() const = 0;
+        virtual std::optional<std::string> fileName() const = 0;
+        virtual std::optional<int> lineNumber() const = 0;
+        virtual std::optional<int> columnNumber() const = 0;
```
How about we combine these into a "SourceSpan" struct? This will help enforce the second part of the invariant through the type system. E.g:
```
struct SourceLoc {
  std::string fileName;
  int lineNum;
  std::optional<int> colNum;
}
struct SourceSpan {
  SourceLoc begin;
  std::optional<SourceLoc> end;
}
```

0xddom replied:
I like the idea but I dont see how a SourceSpan can span across files so I'm going to tweak it a little

Technius replied:
Yup, small brain fart on my part. The file name should go into SourceSpan. Your change LGTM.

Technius reviewed (approved):
LGTM. Some minor nitpicking below

benSepanski commented:
Thanks @Technius and @0xddom! I think we've reached a good point with this PR

Unless there are any objections I'll plan to merge in the next couple hours 🙂
